McCorrections/src/LeptonWeights.cpp

```diff
@@ -65,53 +65,70 @@ LeptonWeights::LeptonWeights(const std::string& electron_idIsoInput, const std::
     muonSF(muon_idIsoInput, muon_SingletriggerInput, muon_CrossTriggerInput), period(_period),
     tau_iso_wp(_tau_iso_wp)
 {
-    if(period == Period::Run2016)
-        tauIdWeight = std::make_shared<TauIDSFTool>("2016Legacy","DeepTau2017v2p1VSjet", ToString(tau_iso_wp), false);
-    else if(period == Period::Run2017)
-        tauIdWeight = std::make_shared<TauIDSFTool>("2017ReReco","DeepTau2017v2p1VSjet", ToString(tau_iso_wp), false);
-    else if(period == Period::Run2018)
-        tauIdWeight = std::make_shared<TauIDSFTool>("2018ReReco","DeepTau2017v2p1VSjet", ToString(tau_iso_wp), false);
-    else
-        throw exception("Period %1% is not supported.") % period;
 
     tauTriggerWeight_eTau = std::make_shared<tau_trigger::SFProvider>(tauTriggerInput, "etau", ToString(tau_iso_wp));
     tauTriggerWeight_muTau = std::make_shared<tau_trigger::SFProvider>(tauTriggerInput, "mutau", ToString(tau_iso_wp));
     tauTriggerWeight_tauTau = std::make_shared<tau_trigger::SFProvider>(tauTriggerInput, "ditau",
                                                                         ToString(tau_iso_wp));
 }
 
-double LeptonWeights::GetIdIsoWeight(EventInfo& eventInfo) const
+std::shared_ptr<TauIDSFTool>& LeptonWeights::GetTauIdProvider(TauIdDiscriminator discr, DiscriminatorWP wp)
 {
-    const ntuple::Event& event = *eventInfo;
-    const Channel channel = static_cast<Channel>(event.channelId);
-    if(channel == Channel::ETau) {
-        double tau_weight = tauIdWeight->getSFvsPT(eventInfo.GetLeg(2).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return electronSF.GetIdIsoSF(eventInfo.GetLeg(1).GetMomentum()) * tau_weight;
+    std::map<Period ,std::string>  period_label = { { Period::Run2016, "2016Legacy" },
+                                                     { Period::Run2017, "2017ReReco" },
+                                                     { Period::Run2018, "2018ReReco" } };
+    std::string s = "by";
+    std::string::size_type i = ToString(discr).find(s);
+
+    tau_sf_providers[discr][wp] = std::make_shared<TauIDSFTool>(period_label.at(period),
+                                ToString(discr).erase(i, s.length()), ToString(wp), false, false);
+    return tau_sf_providers[discr][wp];
+}
+
+double LeptonWeights::GetLegIdIsoWeight(LepCandidate leg, DiscriminatorWP VSe_wp, DiscriminatorWP VSmu_wp,
+                                        DiscriminatorWP VSjet_wp,
+                                        UncertaintySource /*unc_source*/, UncertaintyScale /*unc_scale*/)
+{
+    std::map<LegType, std::pair<TauIdDiscriminator,DiscriminatorWP>> leg_wp_disc = {
+         { LegType::e, { TauIdDiscriminator::byDeepTau2017v2p1VSe, VSe_wp } },
```

Comment:
tau ID discriminators should be applied only to tau leg => this map should not exist